broker/tasks/iam.py

```diff
@@ -135,18 +135,31 @@ def delete_previous_server_certificate(operation_id: str, **kwargs):
     else:
         iam = iam_govcloud
 
-    for certificate in Certificate.query.filter(
+    certificates_to_delete = Certificate.query.filter(
         and_(
             Certificate.service_instance_id == service_instance.id,
             Certificate.id != service_instance.current_certificate_id,
         )
-    ).all():
+    ).all()
+    for certificate in certificates_to_delete:
         try:
+            print(certificate.iam_server_certificate_name)
+            iam.get_server_certificate(
+                ServerCertificateName=certificate.iam_server_certificate_name
+            )
```

Comment:
what's the goal here?

Reply:
It's complicated. And more of a matter of IAM permissions. 

According to AWS support, the only way to make a DeleteServerCertificate operation return NoSuchEntity error instead of AccessDenied is to give iam:DeleteServerCertificate on all certificates (*), rather than certificates under the path managed by this broker. However, these permissions would allow this domain broker to delete certificates managed by other brokers, so we do not want to do that.

Instead, we have updated the delete certificate code to check if the certificate exists before attempting the deletion. If the certificate no longer exists, then we do not attempt the deletion.

Reply:
I guess the real underlying goal was to address errors like this in th elogs

```
external-domain-broker-gov-production is not authorized to perform: iam:DeleteServerCertificate on resource: server certificate <redacted> because no identity-based policy allows the iam:DeleteServerCertificate action, 
```

Reply:
I inverted the control flow so that we only do a `get_server_certificate` call if the `delete_server_certificate` fails with `AccessDenied`, which might happen when the certificate is already deleted. The refactor preserves the existing behavior much better and just catches the undesirable error case

92728c872c559e8658e62743efb9b1f0bc94c556

Reply:
ah, that makes more sense to me now